Increase the system user id range

infinisil opened the pull request:
**This is now a full RFC at NixOS/rfcs#52**

This enlarges the system uid/gid range 6-fold, from 100 to 600 ids. This
is a preventative measure against running out of dynamically allocated
ids for NixOS services with `isSystemUser`, which should become the
preferred way of allocating uids for non-real users if `DynamicUser` isn't applicable.

#### Motivation for this change
We're running out of static uid/gid's in the standard range from 0 to 400, only ~85 are left: https://github.com/NixOS/nixpkgs/blob/9fe0552e4db1253e0a714ed50e288cde217ddab1/nixos/modules/misc/ids.nix#L341-L344

While it might be possible to assign another range for static id's, that's just an ugly solution, and having to declare static id's has always been a bit of a pain.

A better solution is to not assign static id's for services, but to use NixOS dynamic user allocation (happening by default, when `uid/gid = null`). Specifically with `isSystemUser = true` such that ids get allocated in the range `SYS_{U,G}ID_MIN` (400) - `SYS_{U,G}ID_MAX` (499 before this change, 999 after), meaning the range above 1000 can be used exclusively by real users.

Why extend it to 999? Because a limit of 100 ids means that over the lifetime of a NixOS system, users can only ever run 100 different services, since id's aren't repurposed after a service is disabled (See a57bcd38b49bfe9d048b12de3c839bc72b298d2e), which is a bit on the low side. And 999 (=`{U,G}ID_MIN-1`) is also the default value anyways, see `man login.defs`.

This means that instead of having a limit of 400 static uids for *all* NixOS services in nixpkgs, we have a limit of 600 dynamic uids for all NixOS services *ever enable* on a single system.

##### But isn't it bad to not use static id's for services?

I don't think so. One reason I can think of for having static id's is for some service that needs to transfer its files across NixOS machines, but this is not the case for most services. Another reason might be that loading data from backups into a new NixOS installation will make services not work due the mismatch of data and service ids. This is however not a problem if `/var/lib/{u,g}id-map` gets restored as well, and also not a problem if services use `StateDirectory` or `systemd.tmpfiles` which fixes permissions on service start.

See https://github.com/NixOS/nixpkgs/issues/60732#issuecomment-510027542 for a previous discussion

cc @ryantm @arianvp @aanderse @edolstra

infinisil committed:
nixos/users: Increase maximum system uid/gid from 499 to 999

This enlarges the system uid/gid range 6-fold, from 100 to 600 ids. This
is a preventative measure against running out of dynamically allocated
ids for NixOS services with isSystemUser, which should become the
preferred way of allocating uids for non-real users.

Infinisil commented:
While not directly a problem, this is not beneficial for @jameysharp's move to all-static uid's in https://github.com/NixOS/nixpkgs/pull/64594

arcnmx commented:
Is it possible and/or are there plans to unassign the existing ~300 IDs, moving them to either `isSystemUser` or `DynamicUser`? (assuming that the fact they already exist in passwd/uid-map/whatever means they can just remain as-is on existing systems)

Infinisil commented:
I believe it would be safe to switch services that don't need static uid's to `isSystemUser` (because existing systems will get the same id's from `/var/lib/nixos/{u,g}id-map`). `DynamicUser` might also be safe if systemd can handle this transition, this should be tested first.

arcnmx commented:
Last I recall reading about `DynamicUser` it was explicitly stated as a goal that it be able to handle seamless transitions for existing systems (I believe their particular approach is just to treat it as if DynamicUser weren't set if the user already exists on the system?). Certainly yeah stuff to test, just wondering since it seems like it may be unnecessary to keep many of them reserved for the future, which could widen the range of usable system IDs further.

(though systemd is meant to handle the transition, if nixos removes deleted users from passwd that may or may not cause problems; really depends on whether StateDirectory= is the whole story or not for a given service I guess?)

aanderse commented:
Unfortunately moving to `DynamicUser` for many of the existing nixos modules will be difficult because the `dataDir` pattern (or maybe in this case anti-pattern).

reanimus commented:
So, after some discussion in the IRC channel, it looks line roon-server doesn't work with DynamicUser enabled, as the backup mechanism needs to be able to write to a persistent location and DynamicUser would make managing permissions for this directory impossible.

Infinisil commented:
Yeah so it might be the case that we can't use `DynamicUser` as much as we'd like to

volth commented:
> use `StateDirectory` or `systemd.tmpfiles` which fixes permissions on service start.

`chown -R $datadirectories` could be very-very slow if there are many files.
For `hadoop` it is unacceptable slow, could take hours.
Perhaps the same for databases  which could grow big (postgres, arango, etc), I just had no experience to grow them as big as hbase on hadoop.

edolstra commented:
My understanding is that `DynamicUser` is only intended for services that don't have persistent state.

aanderse commented:
@edolstra there is some allowance for persistent data. See 'Persistent Data' section of http://0pointer.net/blog/dynamic-users-with-systemd.html

arianvp commented:
> My understanding is that DynamicUser is only intended for services that don't have persistent state.

Incorrect.  It's only meant for services that don't share state _between_ services. Persistent state is definitely possible, as long as it's defined through `StateDirectory=,CacheDirectory=,LogDirectory=` etc

Infinisil commented:
@volth `StateDirectory` and co. won't chown recursively when the top-most directory already has the correct perms (source: `man systemd.exec`), so this can be used for such cases.

`systemd.tmpfiles` doesn't have such a mechanism unfortunately. It can either chown recursively with `Z` or it doesn't chown at all if that's not used (the `d` directive only does anything if the directory doesn't exist yet) (see `man tmpfiles.d`)

globin commented:
I'd prefer this being a proper RFC :)

Infinisil commented:
I made this a full RFC, see above

asymmetric commented:
AFAICT, systemd ignores `login.Defs` and instead relies on [compilation flags](https://github.com/systemd/systemd/issues/9769#issuecomment-409518586) to decide the uid pool size.

So it seems you'd have to change [this](
https://github.com/NixOs/nixpkgs/tree/c9453e32b0d33a5ec641ef923708a66da3e74001/pkgs/os-specific/linux/systemd/default.nix#L89).

Infinisil commented:
@asymmetric Yeah, but I think this is only relevant for users created through `useradd`, because the dynamic ids allocated by NixOS use the limits in update-users-groups.pl to write /etc/passwd directly

Will do that change too

globin commented:
FYI NixOS/rfcs#52 is in Final Comment Period

Infinisil commented:
@asymmetric Now applied the change to the systemd build flags too. Will probably have to switch to staging because of this.

@GrahamcOfBorg eval

asymmetric commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -83,8 +83,8 @@ stdenv.mkDerivation rec {
     "-Db_pie=true"
-    "-Dsystem-uid-max=499" #TODO: debug why awking around in /etc/login.defs doesn't work
-    "-Dsystem-gid-max=499"
+    "-Dsystem-uid-max=999" #TODO: debug why awking around in /etc/login.defs doesn't work
```
Do you understand what this `TODO` is about?

Infinisil replied:
I kind of took it as "debug why login.defs settings aren't used", but I am a bit confused about the "awking", what does `awk` have to do with this??

asymmetric replied:
Yeah, I also think it’s referring to that. In that case, should we replace the `TODO` with a link to the systemd issue?

Infinisil replied:
Agreed. Apparently there's a couple issues on that: https://github.com/systemd/systemd/issues/3855, https://github.com/systemd/systemd/issues/4850, https://github.com/systemd/systemd/issues/9769, https://github.com/systemd/systemd/issues/9843, https://github.com/systemd/systemd/issues/10184

Guess I'll link to https://github.com/systemd/systemd/issues/9769

asymmetric commented on nixos/modules/programs/shadow.nix:
```diff
@@ -6,17 +6,18 @@ with lib;
 
 let
 
+  # Note that these id ranges are duplicated in update-users-groups.pl
```
Is there a place where we document when `login.defs` gets used, and when `update-users-groups.pl`?

If not, this could be an opportunity to do it. Maybe even just expanding the inline comment a bit could be enough, explaining why the duplication is in place.

Infinisil replied:
Yeah that's a good idea, will do

Infinisil commented:
Turns out that the ranges defined in the systemd compile time options really don't matter that much, they're only used by systemd features like `ConditionUser=@system` and `systemd-sysusers` (which is apparently another way to allocate ids, maybe we could even use that for NixOS at some point).

The ranges in `login.defs` are only used for mutable users/groups added with `useradd`/`groupadd`/`newusers`.

And the ones in `update-users-groups.pl` are used by NixOS to allocate its declaratively defined users.

I added this info to a comment.

Also added links to the issues regarding runtime configuration of systemd ranges.

Infinisil commented:
NixOS/rfcs#52 is now accepted and I think this is ready to be merged.

Infinisil commented:
Hehe, the error on aarch64 is deceivingly related-looking:
```
error: all build users are currently in use; consider creating additional users and adding them to the 'nixbld' group
```

But really that only seems to occur as a result of heavy aarch64 builder usage, nothing with this PR.
